projects/packages/waf/src/class-waf-rules-manager.php

```diff
@@ -263,19 +291,22 @@ public static function generate_automatic_rules() {
 	/**
 	 * Generates the rules.php script
 	 *
-	 * @throws \Exception If filesystem is not available.
-	 * @throws \Exception If file writing fails.
+	 * @global \WP_Filesystem_Base $wp_filesystem WordPress filesystem abstraction.
+	 *
+	 * @throws Waf_Exception         If filesystem is not available.
+	 * @throws File_System_Exception If writing to IP allow list file fails.
+	 * @throws File_System_Exception If writing to IP block list file fails.
+	 *
 	 * @return void
 	 */
 	public static function generate_ip_rules() {
-		/**
-		 * WordPress filesystem abstraction.
-		 *
-		 * @var \WP_Filesystem_Base $wp_filesystem
-		 */
 		global $wp_filesystem;
 
-		Waf_Runner::initialize_filesystem();
+		try {
+			Waf_Runner::initialize_filesystem();
+		} catch ( Waf_Exception $e ) {
+			throw $e;
+		}
```

Comment:
```suggestion
		Waf_Runner::initialize_filesystem();
```

We could also just not have the try-catch block here I think?

Reply:
I was 50/50 as to whether or not I wanted to include these try/catch/re-throw patterns, so I'm happy to proceed without them.

The main reason I like these is that it leads to much more specific PHPdoc `@throws` descriptions. When one method includes a call to another method that potentially throws an exception, it is difficult to document at the parent method level (PHPCS gets angry, plus it wouldn't be ideal/DRY to be documenting one method's exception cases in all of the methods that implement it).

That, and it makes it crystal clear when reading the code where the potential error cases are. By writing the try/catch, we are showing with intention that a `Waf_Exception` could occur at this step, and if so, we want to pass it up.

I'm of course very open for input and opinions regarding this, and happy to make adjustments! 🙇‍♂️😁

```php
/**
 * A function that throws an exception.
 *
 * @throws Exception If not cool enough.
 */
function doSomethingCool() {
    throw new Exception( 'Not cool enough!' );
}

/**
 * A function that calls functions that throw exceptions without catching them.
 */
function doSeveralCoolThingsA() {
    // ...
    doSomethingCool();
    // ...
}

/**
 * A function that calls functions that throw exceptions and catches them.
 * 
 * @throws Exception When something cool fails to happen.
 */
function doSeveralCoolThingsB() {
    // ...
    try {
        doSomethingCool();
    } catch ( Exception $e ) {
        throw $e;
    }
    // ...
}
```

Reply:
Hmm, I see where you're coming from. The way I see it, catching the exceptions there and re-throwing them is just dead code, cause exactly the same thing would happen without doing that 🤷🏻‍♂️ 

And "just" to be explicit about where Exceptions happen, IMHO it is a bit much. (I'd be in favor of removing them.)

About the `@throws` annotations: Do we need these? According to [this doc](https://docs.phpdoc.org/3.0/guide/references/phpdoc/tags/throws.html) they are optional anyway. I don't know if we have a jetpack-wide rule that we must have them or something? IMHO every modern IDE should know where an exception could be thrown or not, making it trivial to know where to catch them or not... :)

Reply:
I'm cool with removing them! https://github.com/Automattic/jetpack/pull/29108/commits/c1d3effd93c8fd23355781bca6b324383c18d498

> IMHO every modern IDE should know where an exception could be thrown or not, making it trivial to know where to catch them or not...

Perhaps the _real_ solution here is for me to switch from VSCode to PHPStorm... 🤔

Reply:
Maybe :) I'd have expected that VScode can do it too though (without knowing much about it).